eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/ImpTest.hs

```diff
@@ -767,13 +784,34 @@ impWitsVKeyNeeded txBody = do
 
 data ImpTestEnv era = ImpTestEnv
   { iteFixup :: Tx era -> ImpTestM era (Tx era)
+  , iteHook ::
+      Either
+        (NonEmpty (PredicateFailure (EraRule "LEDGER" era)))
+        (State (EraRule "LEDGER" era), [Event (EraRule "LEDGER" era)]) ->
+      LedgerEnv era ->
+      LedgerState era ->
+      Tx era ->
+      ImpTestM era ()
```

Comment:
I think the environment and state arguments are unnecessary, because you already have access to those via `getsNES`

Reply:
And the result of running the rule can also be computed inside the hook, but I guess it's faster to use the precomputed result, so it's fine to keep that as an argument

Reply:
By the time this hook is called the state has already been modified.
So, we do need to pass the result and the state to this hook. Environment on the other hand is not necessary, however it is best to pass it as well for consistency.

Moreover, I am leaning towards making it an `Expectation`. This would limit access of this hook to the ImpTest state and protect against mistakes that would affect the outcome of the imp test itself.

Reply:
We need to `translateInputs` within the `iteExpectLedgerRuleConformance` since the translation functionality lives in `cardano-ledger-conformance`. Making the return type of the hook to `Expectation`, my guess is it would require a `liftIO` and the function obtained from `withRunInIO` to be passed to the hook. Please correct me if I'm wrong. I thought it will be a little cumbersome because within `trySubmitTx` we would need to change the context from `ImpTestM era ()` to `IO ()` with `liftIO` to call the hook and within the hook, we would need to change the context back to `ImpTestM (LedgerSpec era) ()` with `runInIO` to call `translateInputs`?